src/analyses/basePriv.ml

```diff
@@ -830,13 +858,35 @@ struct
       vf (V.protected g);
     | _ -> ()
 
-  let invariant_global ask getg g =
+  let invariant_global (ask: Q.ask) getg g =
     let getg = Wrapper.getg ask getg in
     match g with
     | `Left g' -> (* unprotected *)
       ValueDomain.invariant_global (fun g -> getg (V.unprotected g)) g'
-    | `Right g -> (* protected *)
-      Invariant.none
+    | `Right g' -> (* protected *)
+      let locks = ask.f (Q.MustProtectingLocks g') in
+      if LockDomain.MustLockset.is_all locks || LockDomain.MustLockset.is_empty locks then
+        Invariant.none
+      else if VD.equal (getg (V.protected g')) (getg (V.unprotected g')) then
+        Invariant.none (* don't output protected invariant because it's the same as unprotected *)
+      else (
+        let inv = ValueDomain.invariant_global (fun g -> getg (V.protected g)) g' in (* TODO: this takes protected values of everything *)
```

Comment:
The problem here is that it can be unsound I think. If a global pointer is protected but what it points to isn't, then this invariant will assume both are protected.
Surprisingly, Ultimate hasn't found any of such to be wrong. Maybe because if the pointee is jointly protected, then it's fine again?

We cannot just delegate to `read_global` here because this is in flow-sensitive context, so there's no notion of current lockset, etc.

Reply:
I now constructed a regression test for the kind of unsound case I had in mind and provide a fix for it.